daemon/cmd/devices.go

```diff
@@ -0,0 +1,292 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2016-2021 Authors of Cilium
+
+// This module implements Cilium's network device detection.
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	excludedDevicePrefixes = []string{
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+	}
+)
+
+type DeviceManager struct {
+	mu      lock.Mutex
+	devices map[string]bool
+}
+
+func NewDeviceManager() *DeviceManager {
+	return &DeviceManager{
+		devices: make(map[string]bool),
+	}
+}
+
+func (dm *DeviceManager) Close() {
+}
+
+// Detect tries to detect devices which are to be used for:
+// NodePort BPF, direct routing in NodePort BPF, IPsec and Bandwidth Manager.
+//
+// The devices are detected by looking at all the configured global unicast
+// routes in the system.
+func (dm *DeviceManager) Detect() error {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	dm.devices = make(map[string]bool)
+
+	if err := expandDevices(); err != nil {
+		log.WithError(err).Warnf("Failed to expand device wildcards")
+	}
+
+	l3DevOK := false
+	if option.Config.EnableNodePort {
+		// Probe whether fast redirect is supported for L3 devices. This will
+		// invoke bpftool and requires root privileges.
+		l3DevOK = supportL3Dev()
```

Comment:
Had to put the probing behind the EnableNodePort check as otherwise integration tests "Fatal()" when bpftool is missing (or if not run as root). Alternatively could make the probing not fatal on bpftool invocation failure and just return empty set of features, but not sure if that's any nicer.

Reply:
Having bpftool installed is probably better. Will take a look at adding it.

Reply:
Yep I think I meant the privileged unit tests. daemon_test.go I think was stumbling on this.

Reply:
AFAIK, `bpftool` should already be installed in all environments we use for integration tests. Did you mean unit tests maybe?

Reply:
I think the solution as is is fine. The failing tests prior to moving call to supportL3Dev() here was the daemon_test.go which was non-privileged, so we wouldn't be able to query this with bpftool even if it was available.